internal/gatewayapi/listeners.go

```diff
@@ -18,144 +18,190 @@ import (
-
-	var (
-		insecureHostnames = map[string]int{}
-		secureHostnames   = map[string]int{}
-	)
-
-	for _, listener := range listeners {
-		hostname := HostnameDeref(listener.Hostname)
-
-		switch listener.Protocol {
-		case gatewayapi_v1beta1.HTTPProtocolType:
-			// Keep the first insecure listener port we see
-			if result.InsecurePort == 0 {
-				result.InsecurePort = int(listener.Port)
-			}
-
-			// Count hostnames among insecure listeners with the "valid" port.
-			// For other insecure listeners with an "invalid" port, the
-			// "PortUnavailable" reason will take precedence.
-			if int(listener.Port) == result.InsecurePort {
-				insecureHostnames[hostname]++
-			}
-		case gatewayapi_v1beta1.HTTPSProtocolType, gatewayapi_v1beta1.TLSProtocolType:
-			// Keep the first secure listener port we see
-			if result.SecurePort == 0 {
-				result.SecurePort = int(listener.Port)
-			}
-
-			// Count hostnames among secure listeners with the "valid" port.
-			// For other secure listeners with an "invalid" port, the
-			// "PortUnavailable" reason will take precedence.
-			if int(listener.Port) == result.SecurePort {
-				secureHostnames[hostname]++
-			}
-		}
-	}
-
-	for _, listener := range listeners {
-		hostname := HostnameDeref(listener.Hostname)
-
-		if len(hostname) > 0 {
-			if err := IsValidHostname(hostname); err != nil {
+	for i, listener := range listeners {
+		// Check for a valid hostname.
+		if hostname := ref.Val(listener.Hostname, ""); len(hostname) > 0 {
+			if err := IsValidHostname(string(hostname)); err != nil {
 				result.InvalidListenerConditions[listener.Name] = metav1.Condition{
 					Type:    string(gatewayapi_v1beta1.ListenerConditionProgrammed),
 					Status:  metav1.ConditionFalse,
 					Reason:  string(gatewayapi_v1beta1.ListenerReasonInvalid),
 					Message: err.Error(),
 				}
+				continue
 			}
 		}
 
+		// Check for a supported protocol.
 		switch listener.Protocol {
-		case gatewayapi_v1beta1.HTTPProtocolType:
-			switch {
-			case int(listener.Port) != result.InsecurePort:
-				result.InvalidListenerConditions[listener.Name] = metav1.Condition{
-					Type:    string(gatewayapi_v1beta1.ListenerConditionAccepted),
-					Status:  metav1.ConditionFalse,
-					Reason:  string(gatewayapi_v1beta1.ListenerReasonPortUnavailable),
-					Message: "Only one HTTP port is supported",
+		case gatewayapi_v1beta1.HTTPProtocolType, gatewayapi_v1beta1.HTTPSProtocolType, gatewayapi_v1beta1.TLSProtocolType:
+		default:
+			result.InvalidListenerConditions[listener.Name] = metav1.Condition{
+				Type:    string(gatewayapi_v1beta1.ListenerConditionAccepted),
+				Status:  metav1.ConditionFalse,
+				Reason:  string(gatewayapi_v1beta1.ListenerReasonUnsupportedProtocol),
+				Message: fmt.Sprintf("Listener protocol %q is unsupported, must be one of HTTP, HTTPS or TLS", listener.Protocol),
+			}
+			continue
+		}
+
+		// Check for conflicts with other listeners.
+		conflicted := false
+		for j, otherListener := range listeners {
```

Comment:
i think re-looping around all listeners will have us comparing against potentially invalid elements that we could probably ignore and doing a sort of two-way comparison when there are conflicts when we only need to do it once? wdyt about saving the valid listeners (indexed maybe with `envoyListenerName` from below since that ends up capturing protocol/port) in a map and then each listener contains a list of unique hostnames assigned to it, e.g. a structure like

```golang
type ValidateListenersResult struct {
  ListenerNames map[string]string

  // maps name (protocol-port, e.g. http-80 to port details)
  // could maybe just use the Gateway Listener port number
  Ports map[string]ListenerPort

  InvalidListenerConditions map[gatewayapi_v1beta1.SectionName]metav1.Condition
}

type ListenerPort struct {
  port int32
  containerPort int32
  protocol string
  hostnames map[string]struct{}
}
```

this way once a Gateway Listener is completely valid and unique, it is added to the map, and subsequent Listeners are compared against the good ones only, with list order precedence (which is how Routes solve conflicts, not explicitly specified anywhere on Gateways but good for consistency?)

Reply:
I think that's reasonable, will work on refactoring.

Reply:
Started working on this and it seemed easiest to just change the inner loop to only compare a given Listener to other Listeners before it in list order, which has the same logical effect and requires less code shuffling 😁  I also cleaned up a few other things to help readability. LMK if this works for you! https://github.com/projectcontour/contour/pull/5160/commits/6ab84ac2655d6377e2cf9f50a95779717adf2f3a